connectors/llm/interface.py

```diff
@@ -55,30 +64,61 @@ def ask(self, system_prompt, previous_messages, question, agent_id, stream):
                     msg_list.append(AIMessage(content=f"{msg['text']}</s>"))
         prompt.messages = msg_list + prompt.messages
 
-        log.debug("prompt: %s", prompt)
-        model = ChatOpenAI(
+        llm = ChatOpenAI(
             model=cfg.LLM_MODEL_NAME,
             openai_api_base=cfg.LLM_BASE_URL,
             openai_api_key=cfg.LLM_API_KEY,
             temperature=cfg.LLM_TEMPERATURE,
         )
 
-        chain = prompt | model | StrOutputParser()
+        chain = prompt | llm
+
+        log.debug("prompting llm...")
+
+        def get_llm_response():
+            timer_start = None
+
+            with get_openai_callback() as cb:
+                for chunk in chain.stream(prompt_params, stream_usage=True):
+                    if not timer_start:
+                        # this is the first output token received
+                        timer_start = time.time()
+                    if len(chunk.content):
+                        text_content = {"text_content": chunk.content}
+                        yield text_content
+
+                timer_end = time.time()
```

Comment:
There's a slim chance for a bug here. If no chunks are yielded from stream you'll never start the timer and end up calling .time() on None

Reply:
I thought about this too right after I asked for your review :D Good spot